### PR TITLE
Use a custom Value type instead of Maybe for storing rule results

### PR DESCRIPTION
**Pull request by cocreature**
This is a refactoring-only PR in preparation for supporting stale
values in damlc so we can still produce some results if a rule fails
to produce a value but has a stale value stored.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/hie-core/src/Development/IDE/Core/Shake.hs [hunk 1]
```diff
@@ -154,6 +151,25 @@ instance Hashable Key where
 --   not propagate diagnostic errors through multiple phases.
 type IdeResult v = ([FileDiagnostic], Maybe v)
 
+data Value v
+    = Succeeded v
+    | Failed
```
Comment (hurryabit) [inline compiler/hie-core/src/Development/IDE/Core/Shake.hs]: I suppose the plan is to add a third alternative here for stale values?

File: compiler/hie-core/src/Development/IDE/Core/Shake.hs [hunk 2]
```diff
@@ -408,8 +414,9 @@ defineEarlyCutoff op = addBuiltinRule noLint noIdentity $ \(Q (key, file)) old m
         Just res -> return res
         Nothing -> do
             (bs, (diags, res)) <- actionCatch
-                (do v <- op key file; liftIO $ evaluate $ force v) $
+                (do v <- op key file; liftIO $ evaluate $ force $ v) $
```
Comment (hurryabit) [inline compiler/hie-core/src/Development/IDE/Core/Shake.hs]: Somebody's throwing around `$`s. :D

File: compiler/hie-core/src/Development/IDE/Core/Shake.hs (hunk 1 above)
Comment (cocreature) [reply compiler/hie-core/src/Development/IDE/Core/Shake.hs]: Exactly.

File: compiler/hie-core/src/Development/IDE/Core/Shake.hs (hunk 2 above)
Comment (cocreature) [reply compiler/hie-core/src/Development/IDE/Core/Shake.hs]: :money_mouth_face:

**Pull request merged**